fpsd/features.py

```diff
@@ -0,0 +1,765 @@
+                                 ', '.join(crosstab_columns))
+
+        self.execute_query_from_string(query)
+        return "features.cell_ordering"
+
+    def generate_table_outgoing_cell_ordering_differences(self,
+                                                          num_features=500):
+        """This method takes all examples and produces a
+        table with the following format:
+
+        exampleid  | a         | b         ...    | num_features
+        (integer)  | (bigint)  | (bigint)  ...    | (bigint)
+        --------------------------------------------------------
+        9          | 2         | 1         ...
+        10         | 1         | 2         ...
+        11         | 2         | 1         ...
+
+        where there are a variable number (num_features) of
+        feature columns, each named outgoing_cell_position_difference_x
+        where x is the position in the trace + 1, beginning at 1.
+        For example, outgoing_cell_position_difference_1 is the difference in
+        position between the 2nd and 1st outgoing cells.
+
+        Args:
+            num_features [int]: number of features to create
+
+        Returns:
+            [string] name of newly created table
+        """
+
+        num_ranks = num_features + 1
+
+        self.drop_stale_feature_table("features.cell_ordering_differences")
+
+        self._create_temp_packet_positions(outgoing_only=True)
+        self._create_table_outgoing_cell_positions(num_cells=num_ranks)
+
+        self.drop_stale_feature_table("top_{}_cell_ordering".format(num_ranks))
+
+        crosstab_columns = ['outgoing_cell_position_{} bigint'.format(x)
+                            for x in range(1, num_ranks + 1)]
+
+        query = ("CREATE TEMP TABLE top_{n}_cell_ordering AS         "
+                 "(SELECT * FROM crosstab(                           "
+                 "'SELECT * FROM top_{n}_outgoing_packet_positions') "
+                 "AS ct(exampleid integer,                           "
+                 "{cols})); ").format(n=num_ranks,
+                                      cols=', '.join(crosstab_columns))
+        self.execute_query_from_string(query)
+
+        diff_columns = [("({prefix}_{n2} - {prefix}_{n1}) "
+                         "AS {prefix}_difference_{n1}").format(
+                            prefix='outgoing_cell_position', n2=x+1, n1=x)
+                        for x in range(1, num_ranks)]
+
+        feat_columns = ', '.join(diff_columns)
+        query = ("CREATE TABLE features.cell_ordering_differences   "
+                 "AS (SELECT exampleid, {cols}                      "
+                 "FROM top_{n}_cell_ordering); ".format(cols=feat_columns,
+                                                        n=num_ranks))
+
+        self.execute_query_from_string(query)
+        return "features.cell_ordering_differences"
+
+    def generate_table_binned_counts(self, num_features=100, size_window=30):
```

Comment:
Using `num_features` here instead of `num_cells` because each feature does not correspond to an individual cell whereas in the `num_cells` case they do. I’ve tried to make sure that when `num_cells` is used it refers to a feature corresponding to an individual cell.